### PR TITLE
Support two-phase commit of checks (CMC-1235)

### PR DESCRIPTION
**Pull request by itzg**
WIP TODO:

- [x] Unit tests of poller.prepare.* and poller.commit
- [x] Recover lost test coverage
- [x] Update endpoint server to implement abbreviated 2PC

### CODE REVIEW

File: poller/connection.go [hunk 1]
```diff
@@ -30,7 +30,7 @@ import (
 // EleConnection implements Connection
 // See Connection interface for more information
 type EleConnection struct {
-	stream ConnectionStream
```
Comment (itzg) [inline poller/connection.go]: I took some opportunities to eliminate back-references that weren't actually required.

File: poller/connection_stream.go [hunk 2]
```diff
@@ -97,6 +96,25 @@ func (cs *EleConnectionStream) GetConnections() map[string]Connection {
 	return cs.conns
 }
 
+// ReconcileChecks routes the ChecksPreparation to its schedulers.
+func (cs *EleConnectionStream) ReconcileChecks(cp *ChecksPreparation) {
```
Comment (itzg) [inline poller/connection_stream.go]: @rphillips with this interface delegation I was able to theoretically continue to support multi-zone per poller

File: poller/connection_stream_test.go [hunk 3]
```diff
@@ -10,45 +10,6 @@ import (
 	"sync"
 )
 
-func TestConnectionStream_Register(t *testing.T) {
```
Comment (itzg) [inline poller/connection_stream_test.go]: Sorry @dimtruck , had to remove (for now) these old-style check registering tests.

File: poller/connection_whitebox_test.go [hunk 4]
```diff
@@ -1,228 +0,0 @@
-package poller
```
Comment (itzg) [inline poller/connection_whitebox_test.go]: And whitebox testing crumbled with my refactorings...the unfortunate risk of whitebox testing.

File: check/check_http_test.go [hunk 5]
```diff
@@ -99,7 +101,8 @@ func TestHTTPSuccessIncludeBodyAndHeaders(t *testing.T) {
 	  "target_resolver":"",
 	  "disabled":false
 	  }`, ts.URL)
-	check := check.NewCheck(context.Background(), []byte(checkData), func() {})
+	check, err := check.NewCheck(context.Background(), []byte(checkData))
+	require.NoError(t, err)
```
Comment (dimtruck) [inline check/check_http_test.go]: Very nice! Missed this feature

File: check/check_ping_test.go [hunk 6]
```diff
@@ -196,7 +200,8 @@ func TestPingCheck_NoPermissionToPing(t *testing.T) {
 	mock.EXPECT().Statistics().Return(statistics)
 
 	checkData := fmt.Sprintf(checkDataTemplate, count, timeout)
-	c := check.NewCheck(context.Background(), []byte(checkData), func() {})
+	c, err := check.NewCheck(context.Background(), []byte(checkData))
+	require.NoError(t, err)
```
Comment (dimtruck) [inline check/check_ping_test.go]: Seems like we have assertion duplications.  We could combine all the tests to be data driven in one function potentially, making it easier to add new tests (just adding a data object to the array).  Not a big deal, just an observation.

Comment (itzg) [inline check/check_ping_test.go]: Totally agree. I tried to do that with the new scheduler tests. So, yep, we need to circle around and retrofit some of these original tests.

File: poller/checks_preparation.go [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+func (cp *ChecksPreparation) VersionApplies(version int) bool {
+	return cp != nil && cp.TrackingVersion == version
+}
+
+func (cp *ChecksPreparation) IsNewer(version int) bool {
+	return cp == nil || version > cp.TrackingVersion
+}
+
+func (cp *ChecksPreparation) IsOlder(version int) bool {
+	return cp != nil || version < cp.TrackingVersion
```
Comment (dimtruck) [inline poller/checks_preparation.go]: :)

File: poller/checks_preparation_test.go [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+	}
+
+	cp, err := poller.NewChecksPreparation(1, manifest)
+	require.NoError(t, err)
+
+	block := loadTestDataChecks(t, info...)
+	cp.AddDefinitions(block)
+
+	return cp
+}
```
Comment (dimtruck) [inline poller/checks_preparation_test.go]: Same comment re: data driven tests here :).  This is really good stuff!

File: poller/connection_stream_whitebox_test.go [hunk 9]
```diff
@@ -111,20 +112,29 @@ func TestConnection_WaitCh(t *testing.T) {
...
+	  "type":"remote.http",
+	  "timeout":15,
+	  "period":30,
+	  "ip_addresses":{"default":"127.0.0.1"},
+	  "target_alias":"default",
+	  "target_hostname":"",
+	  "target_resolver":"",
+	  "disabled":false
+				}`))
+	require.NoError(t, err)
```
Comment (dimtruck) [inline poller/connection_stream_whitebox_test.go]: Awesome cleanup!  Thank you!

File: poller/connection_whitebox_test.go (hunk 4 above)
Comment (dimtruck) [inline poller/connection_whitebox_test.go]: ;-(

File: poller/checks_preparation.go [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+	check.CheckIn
+
+	Action ActionType
+	// Populated indicates if the check.CheckIn has been fully populated; however,
+	// this is not applicable if Action is protocol.PrepareActionContinue.
+	Populated bool
+}
+
+type ChecksPreparation struct {
+	TrackingVersion int
```
Comment (goru97) [inline poller/checks_preparation.go]: Can use Version instead of TrackingVersion.

Comment (itzg) [inline poller/checks_preparation.go]: I think I was avoiding confusion with the frame's `Version` field, but looking now, I agree I'll change this to `Version`.

File: poller/checks_preparation_test.go (hunk 8 above)
Comment (itzg) [inline poller/checks_preparation_test.go]: @dimtruck most in here are now data driven :)

File: poller/checks_preparation.go [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+	case protocol.PrepareActionRestart:
+		return ActionTypeRestart
+	case protocol.PrepareActionContinue:
+		return ActionTypeContinue
+	default:
+		return ActionTypeUnknown
+	}
+}
+
+func checkPreparationNeedsPopulating(action string) bool {
```
Comment (rphillips) [inline poller/checks_preparation.go]: nit, maybe make this an isCheckPreparationNeedsPopulating

Comment (itzg) [inline poller/checks_preparation.go]: @rphillips how about `doesCheckPreparationNeedPopulating` ?

File: poller/checks_preparation_test.go [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+
+func loadTestDataChecks(t *testing.T, info ...checkLoadInfo) (checks []check.CheckIn) {
+	checks = make([]check.CheckIn, len(info))
+
+	for i, entry := range info {
+		bytes, err := ioutil.ReadFile("testdata/" + entry.name + ".json")
+		require.NoError(t, err)
+
+		json.Unmarshal(bytes, &checks[i])
```
Comment (rphillips) [inline poller/checks_preparation_test.go]: it shouldn't happen, but we should capture the error here, and make sure it doesn't error.

File: poller/connection.go [hunk 13]
```diff
@@ -116,7 +111,7 @@ func (conn *EleConnection) Connect(ctx context.Context, config *config.Config, t
 	}
 	log.Info("  ... Connected")
 	conn.conn = tlsConn
-	conn.session = NewSession(ctx, conn, config)
+	conn.session = NewSession(ctx, conn, conn.checksReconciler, config)
```
Comment (rphillips) [inline poller/connection.go]: hmm. `conn.checkReconciler` seems redundant when conn is getting passed in...

File: poller/checks_preparation_test.go (hunk 12 above)
Comment (itzg) [inline poller/checks_preparation_test.go]: Fixed in next push

File: poller/connection.go (hunk 13 above)
Comment (itzg) [inline poller/connection.go]: It looks that way, but I wanted to eliminate session's awareness of connection's internals. More specifically, I didn't want session to care where the reconciler came from...especially for unit testing.

File: poller/checks_preparation.go (hunk 11 above)
Comment (rphillips) [inline poller/checks_preparation.go]: that works

**Comment by itzg** [comment]
@rphillips this is ready for one last review. Includes prepare-commit support in the endpoint (and some ripple)

File: endpoint/agent.go [hunk 14]
```diff
@@ -0,0 +1,364 @@
...
+		log.WithFields(log.Fields{
+			"agent":   a,
+			"pending": ourPending,
+		}).Debug("Committing checks to agent")
+		a.sendTo(msgId, protocol.MethodPollerCommit, prepareCommitParams)
+
+	default:
+
+		a.checksLock.Lock()
+		log.WithFields(log.Fields{
```
Comment (rphillips) [inline endpoint/agent.go]: we may want to take the log out of the critical section, for performance.

**Comment by rphillips** [comment]
@itzg really nice job.. I saw a few places where there is a log statement within critical sections. I suggest moving those outside those sections. Otherwise, +1

**Pull request merged**